packages/shorebird_cli/lib/src/flutter_artifacts.dart

```diff
@@ -0,0 +1,124 @@
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:scoped/scoped.dart';
+import 'package:shorebird_cli/src/process.dart';
+import 'package:shorebird_cli/src/shorebird_env.dart';
+
+/// All Flutter artifacts used explicitly by Shorebird.
+enum FlutterArtifact {
+  /// The gen_snapshot executable.
+  genSnapshot,
+
+  /// The analyze_snapshot executable.
+  analyzeSnapshot,
+}
+
+/// A reference to a [FlutterArtifacts] instance.
+final flutterArtifactsRef = create<FlutterArtifacts>(
+  () => const FlutterCachedArtifacts(),
+);
+
+/// The [FlutterArtifacts] instance available in the current zone.
+FlutterArtifacts get flutterArtifacts => read(flutterArtifactsRef);
+
+/// {@template flutter_artifacts}
+/// A class that provides access to Flutter artifacts.
+/// {@endtemplate}
+abstract class FlutterArtifacts {
+  /// {@macro flutter_artifacts}
+  const FlutterArtifacts();
+
+  /// Returns the path to the given [artifact].
+  String getArtifactPath({required FlutterArtifact artifact});
+}
+
+/// {@template flutter_cached_artifacts}
+/// A class that provides access to cached Flutter artifacts.
+/// {@endtemplate}
+class FlutterCachedArtifacts implements FlutterArtifacts {
+  /// {@macro flutter_cached_artifacts}
+  const FlutterCachedArtifacts();
+
+  @override
+  String getArtifactPath({
+    required FlutterArtifact artifact,
+  }) {
+    switch (artifact) {
+      case FlutterArtifact.genSnapshot:
+        return _genSnapshotFile.path;
+      case FlutterArtifact.analyzeSnapshot:
+        return _analyzeSnapshotFile.path;
+    }
+  }
+
+  File get _genSnapshotFile {
+    return File(
+      p.join(
+        shorebirdEnv.flutterDirectory.path,
+        'bin',
+        'cache',
+        'artifacts',
+        'engine',
+        'ios-release',
```

Comment:
This section of directories is repeated – might be nice to extract to a shared place. The executable names could also be made a property of the `FlutterArtifact` enum.

Reply:
I didn't want to over-optimize here because I wanted to keep the local engine artifacts completely decoupled from the flutter cached artifacts since technically there's no reason they need to have the same path segments and/or executable names 🤷

Reply:
Will merge as is for now but lmk if you feel strongly and I'm happy to refactor in a subsequent PR

Reply:
Nope, just an idea for cleanup. Not hurting anything as it is :)